lib/Doctrine/DBAL/Platforms/AbstractPlatform.php

```diff
@@ -3605,4 +3611,9 @@ protected function getLikeWildcardCharacters() : string
     {
         return '%_';
     }
+
+    protected function getDropPrimaryKeySQL(string $table) : string
+    {
+        return 'ALTER TABLE ' . $table . ' DROP PRIMARY KEY';
```

Comment:
I see, `getDropPrimaryKeySQL()` is being introduced here for an abstract platform but it's only needed for MySQL and has MySQL-specific syntax. This is a no-go.

I believe, the original issue is filed for MySQL where `ALTER TABLE DROP INDEX PRIMARY` is not supported. Instead of introducing this method, `MySqlPlatform::getDropIndexSQL()` should be updated to produce correct SQL for the PK.

Reply:
The issue originally discovered during work with Postgres. Initially I tried to modify `getDropIndexSQL()` method but it can't be done for postgres at least, so separate method for primary keys only is the best solution.

Reply:
PK is not just an index, but it's a constraint in some RDBMS. Handling could be done by changing `getDropIndexSQL()` signature but I believe it's not the best option.

Reply:
> so separate method for primary keys only is the best solution

It's fine to have it for Postgres and MySQL which seem unable to drop the PK by name but it shouldn't be part of the abstract platform API. Those platforms which can drop PK by its name shouldn't have any changes.

Reply:
I really don't understand what you propose. Do you mean define `getDropPrimaryKeySQL` as:
```
    protected function getDropPrimaryKeySQL(string $table) : string
    {
        $this->getDropIndexSQL('primary', $table);
    }
```
?
I can't get rid of `getDropPrimaryKeySQL` in favour  `getDropIndexSQL()` because `$index` defined as `Index|string`. So, BC break required, if you would like to use only `getDropIndexSQL()`.

Reply:
> I really don't understand what you propose. Do you mean define `getDropPrimaryKeySQL` as:
> ```php
> protected function getDropPrimaryKeySQL(string $table) : string
> {
>     $this->getDropIndexSQL('primary', $table);
> }
> ```

No. I propose to get rid of this method entirely. The fact that it's introduced in `AbstractPlatform` with a MySQL-specific implementation and is called from `getPreAlterTableIndexForeignKeySQL()` is itself a BC break. If a platform like SQL Server doesn't support the removal of a PK, it will have to implement a workaround like you did, otherwise it will not work anymore.

Instead, for the platforms which cannot drop a PK by name, I propose an override like this:
```php
function getDropIndexSQL($index)
{
    if ($index->isPrimary()) {
       return $theProperSQLToDropIt;
    }

    return parent::getDropIndexSQL($index);
}

Reply:
Hi @morozov ,

I still don't understand why introduction of new `protected` method is worse than changing signature of existing `public` method. Overriding like you proposed, was the very first way I used to fix this issue.

Any other contributors opinions?

Reply:
Regarding platform-specific code - I can declare `getDropPrimaryKeySQL()` as `abstract` and implement it in each inherited class.

Reply:
@SCIF why does the signature of the public method should be changed? If you're referring to my example, it's just pseudo-code.

Declaring the method as abstract is still the same BC break with the same problem: all platforms (including the ones maintained outside of this repo) will have to implement it sometimes at the cost of a workaround.

Instead of dragging all platforms into fixing a platform-specific issue (MySQL and PosrgreSQL, as far as I understand), it should be fixed in those platforms w/o introducing a new API.

Reply:
@morozov , here is [short explanation](https://github.com/doctrine/dbal/pull/2929#discussion_r178455710). To be able to introduce platform specific workaround I need to check if index is primary in `getDropIndexSQL()` but method signature allows passing index as a string (just name of index). In several cases, name of index supplied to `getDropIndexSQL()` is not precise i.e. supplied `primary` rather than actual name.

Reply:
If my current change will be marked as breaking BC, then I would prefer to change signature of `getDropIndexSQL()` to enforce supplying `Index` only, no string. It will bring much more beauty diff`s. Call to BC guru @Ocramius.

Reply:
I've estimated complexity of changing signature of `getDropIndexSQL()`  and it will cause changing signature of `dropIndex()` as well, which is public API.

Reply:
It doesn't require any signature change in `getDropIndexSQL()`. For Postgres:
```php
public function getDropIndexSQL($index, $table = null)
{
    // after $index contains the index name
    if ($index === 'primary') {
        return $this->getDropConstraintSQL($table . '_pkey', $table);
    }

    return 'DROP INDEX ' . $index;
}
```
Something similar for MySQL.

Reply:
Then, in the case when `$index` is a string, you get its object representation and see if it's primary or not.

Reply:
You also can fix it only for your use case when the `$index` comes as an object from the schema manager. If there's another case when dropping the PK by name is required, it can be solved separately.

Reply:
@morozov 

> Then, in the case when $index is a string, you get its object representation and see if it's primary or not.

How I can do that? I don't know a way to do it. If you mean to find a workaround for particular `getAlterTableSQL()` usage — yes, I can fix it in another way, but it's bloody stupid idea to find a big issue and solve only one of use-cases rather than fix it in principal. I've spent a lot of time on supporting this PR and would like to solve a root of evil, rather than just superficial solution of the problem.

Reply:
> How I can do that? I don't know a way to do it.

Neither do I. That's why I'm saying it's fine if you solve only the case where the passed index is an object.

> I've spent a lot of time on supporting this PR and would like to solve a root of evil, rather than just superficial solution of the problem.

The root can could be only solved by reworking the entire index-related API (I recall there was a discussion or at least mention of that /cc @deeky666). This API would account for all differences between indices, constraints, primary keys, foreign keys, etc. in different platforms and would provide some unified higher level API.

It's impossible to solve it correctly in terms of the current API without introducing more controversial methods (I'm talking about `getDropPrimaryKeySQL`()). You can either solve your case in 2.x or postpone it until 3.x and solve properly.

Reply:
Right now, `Index` class contain primary so, really not necessary to be revised because reflect it's purpose pretty fine. The problem is that index can be supplied by name, which would be my proposal to drop in 3.x.
@morozov , thanks for your input. I will refactor this PR just to fix comparing issue soon.